apps/web/src/components/settings/PushNotifications/hooks/useNotificationsRenewal.ts

```diff
@@ -0,0 +1,169 @@
+    }
+
+    // Determine the Safes that need to be renewed for all chains
+    const safesForRenewal = Object.values(allPreferences).reduce<NotifiableSafes>(
+      (acc, { chainId, safeAddress }) =>
+        checkIsRenewalNeeded(chainId, safeAddress)
+          ? { ...acc, [chainId]: [...(acc[chainId] || []), safeAddress] }
+          : acc,
+      {},
+    )
+
+    return isEmpty(safesForRenewal) ? undefined : safesForRenewal
+  }, [
+    safeLoaded,
+    safe.chainId,
+    allPreferences,
+    getChainPreferences,
+    checkIsRenewalNeeded,
+    isNotificationsRenewalEnabled,
+  ])
+
+  // Number of Safes that need to be renewed for notifications
+  const numberSafesForRenewal = useMemo(() => {
+    return safesForRenewal ? flatten(Object.values(safesForRenewal)).length : 0
+  }, [safesForRenewal])
+
+  // Number of chains with Safes that need to be renewed for notifications
+  const numberChainsForRenewal = useMemo(() => {
+    return safesForRenewal ? Object.values(safesForRenewal).filter((addresses) => addresses.length > 0).length : 0
+  }, [safesForRenewal])
+
+  // Boolean indicating if a notifications renewal is needed for any Safe
+  const needsRenewal = useMemo(() => {
+    if (safeLoaded) {
+      return safesForRenewal?.[safe.chainId]?.includes(safe.address.value) || false
+    }
+    return numberSafesForRenewal > 0
+  }, [numberSafesForRenewal, safe.address.value, safe.chainId, safeLoaded, safesForRenewal])
+
+  /**
+   * Function to renew the notifications for the Safes that need it
+   * @returns a Promise that resolves when the notifications have been renewed
+   */
+  const renewNotifications = useCallback(async () => {
+    if (safesForRenewal) {
+      return registerNotifications(safesForRenewal)
+    }
+  }, [safesForRenewal, registerNotifications])
+
+  useEffect(() => {
+    if (
+      shouldShowRenewalNotification &&
+      !!wallet &&
+      !!preferences &&
+      safeLoaded &&
+      !isWrongChain &&
+      !safeTokenVersion &&
+      !hasNotificationMessage &&
+      isNotificationsRenewalEnabled
+    ) {
+      dispatch(
+        showNotification({
+          message:
+            'We’ve upgraded your notification experience. Sign this message to keep receiving important updates seamlessly.',
+          variant: 'warning',
+          groupKey: 'renewal',
+          link: {
+            onClick: () => renewNotifications(),
+            title: 'Sign',
+          },
+        }),
+      )
+
+      // Set the token version to V1 to avoid showing the notification again
+      setTokenVersion(NotificationsTokenVersion.V1)
+    }
+  }, [
+    dispatch,
+    renewNotifications,
+    shouldShowRenewalNotification,
+    preferences,
+    safeLoaded,
+    safe,
+    safeTokenVersion,
+    isWrongChain,
+    hasNotificationMessage,
+    wallet,
+    setTokenVersion,
+    isNotificationsRenewalEnabled,
+  ])
+
+  return { safesForRenewal, numberChainsForRenewal, numberSafesForRenewal, renewNotifications, needsRenewal }
+}
```

Comment:
This hook is very big, any way it can be broken down into simpler functions? E.g. extract some helper functions?

Reply:
I extracted the logic that shows a renewal notification into a separate hook called `useShowNotificationsRenewalMessage`